domain/annotation/state/state.go

```diff
@@ -0,0 +1,274 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/annotations"
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain"
+)
+
+// State represents a type for interacting with the underlying state.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new State for interacting with the underlying state.
+func NewState(factory database.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// GetAnnotations will retrieve all the annotations associated with the given ID from the database.
+// If no annotations are found, an empty map is returned.
+func (st *State) GetAnnotations(ctx context.Context, id annotations.ID) (map[string]string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Prepare query for getting the UUID of ID
+	// No need if kind is model, because we keep annotations per model in the DB.
+	var kindQueryStmt *sqlair.Statement
+	var kindQuery string
+	var kindQueryParam sqlair.M
+
+	if id.Kind != annotations.KindModel {
+		kindQuery, kindQueryParam, err = uuidQueryForID(id)
+		if err != nil {
+			return nil, errors.Annotatef(err, "preparing get annotations query for ID: %q", id.Name)
+		}
+		kindQueryStmt, err = sqlair.Prepare(kindQuery, sqlair.M{})
+		if err != nil {
+			return nil, errors.Annotatef(err, "preparing get annotations query for ID: %q", id.Name)
+		}
+	}
```

Comment:
Have a `getAnnotationsForModel` and `getAnnotations`. It's better to split the logic up front then to try and work out what's going on in the method. It's overly complicated.